idea/src/org/jetbrains/kotlin/idea/inspections/RedundantCompanionReferenceInspection.kt

```diff
@@ -29,6 +32,21 @@ class RedundantCompanionReferenceInspection : AbstractKotlinInspection() {
             if (!objectDeclaration.isCompanion()) return
             if (expression.text != objectDeclaration.name) return
 
+            val property = parent.getStrictParentOfType<KtProperty>()
+            if (property != null && property.name == parent.selectorExpression?.text) return
```

Comment:
What about case like below? Here we still cannot delete `Companion`, but your check does not work here.

```
class My {
    val name = "My name"
    val greeting = "Hello " + Companion.name
    companion object {
        val name = "Your name"
    }
}
```